client/tinychain/ml/linalg.py

```diff
@@ -273,3 +277,317 @@ def svd(matrix: Tensor) -> Tuple:
+                        Map(j=j + 1, n=txn.n, l=l, i=i)
+                    ))
+
+                @closure(txn.n)
+                @post_op
+                def cond(j: UInt) -> Bool:
+                    return j < txn.n
+
+                loop = After(V[l:,i].write((U[i,l:] / U[i,l]) / g), While(cond, step, Map(j=l, n=txn.n, l=l, i=i)))
+
+                return loop
+
+            cxt.update_cols = update_cols
+            cxt.res = cxt.update_cols(n=txn.n, l=l, i=i)
+            return If((g != 0.0), cxt.res)
+
+        cxt.update_g = Float(e[i])
+        cxt.update_cols_wo_last = update_cols_wo_last
+        cxt.update_cols_funct = After(cxt.update_cols_wo_last(l=l, g=g), [V[i,l:].write(0.0), V[l:,i].write(0.0)])
+        
+        cxt.check_last_col = If(
+            (i < txn.n-1),
+            cxt.update_cols_funct)
+        
+        g_final = After([cxt.check_last_col, V[i,i].write(1.0)], cxt.update_g)
+
+        return Map(i=i - 1, l=i, g=g_final, U=U, V=V, e=e)
+
+    @post_op
+    def cond(i: UInt) -> Bool:
+        return i >= 0
+
+    return Map(While(cond, step, Map(i=txn.n-1, l=0, g=0.0, U=U, V=V, e=e)))
+
+
+# TODO: make it work like numpy version
+@post_op
+def lht(txn, U: Tensor, W: Tensor) -> Map:
+    """Accumulation of left hand transformations (lht)"""
+
+    txn.shape = U.shape
+    txn.m, txn.n = [UInt(dim) for dim in txn.shape.unpack(2)]
+    txn.min_m_n = UInt(If(txn.m < txn.n, txn.m, txn.n))
+    
+    @closure(U, W, txn.m, txn.n)
+    @post_op
+    def step(cxt, i: UInt, l: UInt, g: F32) -> Map:
+        
+        @closure(U, W, txn.m, txn.n)
+        @post_op
+        def update_cols(cxt, i: UInt, l: UInt, g: F32) -> F32:
+
+            @closure(U, W, txn.m, txn.n)
+            @post_op
+            def step(j: UInt, l: UInt, i: UInt) -> Map:
+                cxt.col_j = U[i:,j]
+                cxt.update_col = cxt.col_j.write(cxt.col_j + U[i:, i] * ((dot(U[l:, i], U[l:, j])/U[i,i])*g))
+                return Map(After(
+                    cxt.update_col,
+                    Map(j=j + 1, n=txn.n, l=l, i=i)
+                ))
+
+            @closure(txn.n)
+            @post_op
+            def cond(j: UInt) -> Bool:
+                return j < txn.n
+
+            cxt.update_g = 1.0 / g
+            loop = After(cxt.update_g, While(cond, step, Map(j=l, n=txn.n, l=l, i=i)))
+            update_U = After([loop, U[i:,i].write(U[i:,i]*cxt.update_g)], cxt.update_g)
+
+            return If((g != 0.0), update_U, g)
+
+        cxt.update_cols = update_cols
+        cxt.new_l = UInt(i + 1)
+        cxt.l = UInt(After(U[i, cxt.new_l:].write(0.0), cxt.new_l))
+        cxt.new_g = Float(W[i])
+        cxt.final_g = cxt.update_cols(i=i, l=cxt.l, g=cxt.new_g)
+
+
+        return Map(i=i - 1, l=cxt.l, g=cxt.final_g, U=U, W=W)
+
+    @post_op
+    def cond(i: UInt) -> Bool:
+        return i >= 0
+
+    return Map(While(cond, step, Map(i=txn.min_m_n-1, l=0, g=0.0, U=U, W=W)))
+
+
+# helper functions
+def sqr(a):
```

Comment:
I don't think this is any more efficient than just calling `0.0**2`--was there some other reason you needed this? Is it more numerically stable?